docs/docs/tutorials/first-agent/4-human-in-the-loop.md

```diff
@@ -0,0 +1,331 @@
+
+```
+User: How's the weather today in sf?
+Agent:
+```
+
+Let's update how we log the output of the agent so we can see what's going on. Find the final `console.log` and update it to remove the final `.content` so it matches the following:
+
+```ts
+console.log("Agent: ", output.messages[output.messages.length - 1]);
+```
+
+Now when the agent wants to use a tool, we can see the full request. Try it out by running `npx tsx chatloop.ts`. Your result should look something like this:
+
+```
+User: I'm learning LangGraph. Could you do some research on it for me?
+Agent:  AIMessage {
+  "id": "chatcmpl-A4ZQ4tc5ILjuPX8oV0Ovud2W7pqpr",
+  "content": "",
+  "response_metadata": {
+    "tokenUsage": {
+      "completionTokens": 19,
+      "promptTokens": 87,
+      "totalTokens": 106
+    },
+    "finish_reason": "tool_calls",
+    "system_fingerprint": "fp_483d39d857"
+  },
+  "tool_calls": [
+    {
+      "name": "tavily_search_results_json",
+      "args": {
+        "input": "LangGraph"
+      },
+      "type": "tool_call",
+      "id": "call_CRsquDkg5zJ5DGwSqXCA7KP5"
+    }
+  ],
+  "invalid_tool_calls": [],
+  "usage_metadata": {
+    "input_tokens": 87,
+    "output_tokens": 19,
+    "total_tokens": 106
+  }
+}
+```
+
+If you try to continue the conversation, you'll get an error because it's expecting the next message to come from a tool, not a human. We'll fix that in the next step.
+
+## Step 2: Add human confirmation
+
+Right now, the code in our chat loop that is responsible for running the agent and printing the result looks something like this:
+
+```ts
+// Run the chatbot and add its response to the conversation history
+const output = await app.invoke(
+	{
+		messages: [{ content: answer, role: "user" }]
+	},
+	{ configurable: { thread_id: "42" } }
+);
+
+console.log("Agent: ", output.messages[output.messages.length - 1]);
+```
+
+There's nothing here to detect if the agent is trying to run a tool, nor is there a way for a human to confirm that it's okay for the tool to run. We need to change that!
+
+Notice that in the `AIMessage` object example at the end of step 1, the `AIMessage` object has a `tool_calls` field. If that field contains an array, the agent is requesting a tool run. Otherwise, the field will be `undefined`. Let's update our chat loop to check for it and ask the human if graph execution should continue.
+
+Add the following code in between the `app.invoke()` call and the subsequent `console.log()` that prints the output:
+
+```ts
+// 1. Check if the AI is trying to use a tool
+const lastMessage = output.messages[output.messages.length - 1];
+if (lastMessage instanceof AIMessage && lastMessage.tool_calls !== undefined) {
```

Comment:
AIMessage requires a new import at the top:

`import { AIMessage } from "@langchain/core/messages";`